appinventor/components/src/com/google/appinventor/components/runtime/CSVFile.java

```diff
@@ -0,0 +1,287 @@
+            rows = CsvUtil.fromCsvTable(result);
+
+            // Construct column lists from rows
+            constructColumnsFromRows();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Rows property getter method
+     *
+     * @return a YailList representing the parsed rows of the CSV file.
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Returns a list of rows corresponding" +
+            " to the CSV file's content.")
+    public YailList Rows() {
+        return rows;
+    }
+
+
+    /**
+     * Columns property getter method
+     *
+     * @return a YailList representing the parsed columns of the CSV file.
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Returns a list of columns corresponding" +
+            " to the CSV file's content.")
+    public YailList Columns() {
+        return columns;
+    }
+
+    /**
+     * Column names property getter method.
+     * The intended use case of the method is for CSV files which contain
+     * the column names in the first row.
+     *
+     * @return  a YailList containing the elements of the first row.
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Returns the elements of the first row" +
+            " of the CSV contents.")
+    public YailList ColumnNames() {
+        return columnNames;
+    }
+
+    /**
+     * Sets the source file to parse CSV from, and then parses the CSV
+     * file asynchronously.
+     * The results are stored in the Columns, Rows and ColumnNames properties.
+     *
+     * @param source  Source file name
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Indicates source file to load data from." +
+            "Prefix the filename with / to read from a specific file on the SD card. " +
+            "for instance /myFile.txt will read the file /sdcard/myFile.txt. To read " +
+            "assets packaged with an application (also works for the Companion) start " +
+            "the filename with // (two slashes). If a filename does not start with a " +
+            "slash, it will be read from the applications private storage (for packaged " +
+            "apps) and from /sdcard/AppInventor/data for the Companion.")
+    @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_ASSET,
+        defaultValue = "")
+    public void SourceFile(String source) {
+        this.sourceFile = source;
+
+        // Parse CSV after setting source
+        parseCSVFromSource(sourceFile);
+    }
+
+    /**
+     * Waits until CSV Parsing is done by blocking the calling thread.
+     *
+     * The method should be called asynchronously to prevent freezing of
+     * the main thread.
+     *
+     * TODO: This could return a value instead to prevent some async ordering issues.
```

Comment:
Add TODO as an issue, or fix it in this PR.

Reply:
To be tested!

Reply:
The currently chosen 2 async thread approach will not queue tasks properly (e.g. there can be a case where a component is waiting on a CSVFile, and then something like the Clear block could be invoked and thus the value are not wiped properly)

Reply:
Possibly making use of Future objects will solve this?